talk/basicconcepts/assert.tex

```diff
@@ -0,0 +1,89 @@
+\subsection[assert]{Assertions}
+
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Assertions}
+  \begin{block}{Checking invariants in a program}
+    \begin{itemize}
+      \item An invariant is a property of a program that is always true, and the program might crash or yield wrong results if violated
+      \begin{itemize}
+        \item ``Here, `a' should always be positive''
```

Comment:
Isn't this a contradiction? An invariant is supposed to be true during the entire duration of a program. But the quote says: `Here, ...`.

I am also unsure whether the term invariant is actually correct. An assertion tests a condition at a specific point in program execution. The condition may change immediately after the assertion or have been different before.

Reply:
I wouldn't say it is. I don't have an authoritative academic source at hand, but e.g. on [wikipedia](https://en.wikipedia.org/wiki/Invariant_(mathematics)#Invariants_in_computer_science), emphasis mine:
> In [computer science](https://en.wikipedia.org/wiki/Computer_science), an invariant is a [logical assertion](https://en.wikipedia.org/wiki/Logical_assertion) **that is always held to be true during a certain phase of execution of a [computer program](https://en.wikipedia.org/wiki/Computer_program)**. For example, a [loop invariant](https://en.wikipedia.org/wiki/Loop_invariant) is a condition that is true at the beginning and the end of every execution of a loop.
>
> [...]
>
> Programmers often use [assertions](https://en.wikipedia.org/wiki/Assertion_(computing)) in their code to make invariants explicit. Some [object oriented](https://en.wikipedia.org/wiki/Object_oriented) [programming languages](https://en.wikipedia.org/wiki/Programming_language) have a special syntax for specifying [class invariants](https://en.wikipedia.org/wiki/Class_invariant).